docs/docs/async-dialogs.md

```diff
@@ -4,3 +4,628 @@ title: Async Dialogs
+  $('#okButton').on('click.promptDialog', function() {
+    resolve(input.value);
+  });
+  $('#cancelButton').on('click.promptDialog', reject);
+})
+.finally(function() {
+  $('#okButton').off('click.promptDialog');
+  $('#cancelButton').off('click.promptDialog');
+});
+```
+
+Now looking at the example above there are a few issues with it. Mainly the need
+for the deferred anti-pattern and it feels like it is doing too much. A *squint*
+test reveals four inner functions, six variable references, and its behavior is
+to show the dialog, attach two DOM events, construct a promise, hide the dialog,
+and finally detach DOM events. That is a lot for one little function. A
+refactoring can help.
+
+Abstraction is the key here. We will make an *object* (or class) that is
+responsible for managing the dialog box. Its interface will manage only two
+function references (callbacks): when the user clicks ok and when user clicks
+cancel. And it will offer the value when asked.
+
+Using an abstraction like this the `promisePrompt` no longer needs to know
+anything about the DOM and concentrates on just providing a promise. This will
+also make things easier to create a promised version of a progress bar or
+confirmation dialog or any other type of UI that we want to have a value for.
+All we will need to do is write a class for that dialog type with the same
+interface and just pass that class into our promise making method.
+
+The dialog interface might look like this:
+
+```javascript
+var noop = function() {
+  return this;
+};
+
+function Dialog() {
+  this.setCallbacks(noop, noop);
+}
+Dialog.prototype.setCallbacks = function(okCallback, cancelCallback) {
+  this._okCallback     = okCallback;
+  this._cancelCallback = cancelCallback;
+  return this;
+};
+Dialog.prototype.waitForUser = function() {
+  var _this = this;
+  return new Promise(function(resolve, reject) {
+    _this.setCallbacks(resolve, reject);
+  });
+};
+Dialog.prototype.show = noop;
+Dialog.prototype.hide = noop;
+```
+
+Initially the Dialog class sets the two callbacks to *noop* functions. It is up
+to the child class to call them when necessary. We break down the promise
+creation to one function `waitForUser()` that sets the callbacks and returns a
+promise. At this level the `show()` and `hide()` are just *noop* functions as
+well and will be implemented by the child classes.
+
+Our `PromptDialog` class is responsible for inheriting from `Dialog` and setting
+up the required DOM scaffolding and eventually call `this._okCallback` or
+`this._cancelCallback` as appropriate.
+
+It might look like this:
+
+```javascript
+function PromptDialog() {
+  Dialog.call(this);
+  this.el           = document.getElementById('dialog');
+  this.inputEl      = this.el.querySelector('input');
+  this.messageEl    = this.el.querySelector('.message');
+  this.okButton     = this.el.querySelector('button.ok');
+  this.cancelButton = this.el.querySelector('button.cancel');
```

Comment:
There are no non-terrible patterns when you're dealing directly with the DOM, and this one isn't particularly bad - if just `this.el` was injected in the constructor, I think it would be okay...